### ISSUE TITLE
Make it possible to use as a expo module

### ISSUE THREAD

**Issue by axelhildingson** [Adyen/adyen-react-native#108]
Since a lot of new react native project is built upon expo and expo moving over to modules based infrastructure  where developer can build integrate their own native code rather then be stuck with expos pre-configured models. 

So it would be great since this is a fairly new project and stil in beta to trying to fit it into a expo module so a lot more project can pick it up and start using it with go through the hassle of setting it up.

There is a well written documentation on expos website how to build a configure a model
https://docs.expo.dev/modules/overview/

**Comment by descorp** [comment]
Hey @axelhildingson 

Thank you for a suggestion!
We are indeed have expo integration in our backlog 💚

**Comment by szymonzmyslony** [comment]
Hey,
how's it going on this ?
Do you guys have a roadmap ? 
Can we start actively contributing ?

**Comment by descorp** [comment]
Hey @szymonzmyslony 

> Can we start actively contributing ?

Sure! We are always open for contribution 💚

> how's it going on this ? 

Problem we see right now:
there are several [installation steps](https://docs.adyen.com/online-payments/react-native/drop-in#2-install) necessary for our Drop-in to work; we do not have full understanding if it is possible to automate this for Expo.

> Do you guys have a roadmap ?

Short summary of our roadmap is [summarised here](https://github.com/Adyen/adyen-react-native/discussions/136)
Expo is on our backlog, but we did not conducted any research / major steps just yet.

**Comment by anthlasserre** [comment]
Hey all 👋🏼 
I've already spent some time exploring that possibility. Unfortunately, we can't support Adyen SDK into Expo Go without making a dedicated **[Expo Library](https://docs.expo.dev/modules/native-module-tutorial/)**. I've never done that before.

But I found a way to make it works with Development Build at least. And I've automated the necessary installation steps with some Expo Config Plugins.

I'd be happy if anyone can give me some advice to go further.

1. Convert your `app.json` into `app.config.ts` 
2. Paste these Expo Config Plugins functions
```typescript
const config = {
  ...
  name: 'my-app',
  ...
};

const withAdyenAndroid: ConfigPlugin = config => {
  const configWithMainActivity = withMainActivity(config, async config => {
    const mainActivity = config.modResults;
    mainActivity.contents = mainActivity.contents.replace(
      'import com.facebook.react.ReactRootView;\n',
      'import com.facebook.react.ReactRootView;\nimport com.adyenreactnativesdk.AdyenCheckout;\n'
    );
    mainActivity.contents = mainActivity.contents.replace(
      'super.onCreate(null);\n  }',
      'super.onCreate(null);\n    AdyenCheckout.setLauncherActivity(this);\n  }'
    );
    return config;
  });

  const configWithManifest = withAndroidManifest(
    configWithMainActivity,
    async config => {
      const mainActivity = config.modResults;
      // Add com.adyenreactnativesdk.component.dropin.AdyenCheckoutService service
      // after com.facebook.react.HeadlessJsTaskService
      mainActivity.manifest.application = [
        // @ts-expect-error - manifest is not well typed
        {
          ...mainActivity.manifest.application?.[0],
          service: [
            {
              $: {
                'android:name':
                  'com.adyenreactnativesdk.component.dropin.AdyenCheckoutService',
                'android:exported': 'false',
              },
            },
          ],
        },
      ];
      return config;
    }
  );

  return configWithManifest;
};

const withAdyenIos = (config: ExpoConfig) => {
  const appDelegate = withAppDelegate(config, async config => {
    const appDelegate = config.modResults;
    appDelegate.contents = appDelegate.contents.replace(
      '\n\n#if RCT_NEW_ARCH_ENABLED',
      '\n#import <adyen-react-native/ADYRedirectComponent.h>\n\n#if RCT_NEW_ARCH_ENABLED'
    );
    appDelegate.contents = appDelegate.contents.replace(
      'return [super application:application openURL:url options:options] || [RCTLinkingManager application:application openURL:url options:options];',
      'return [ADYRedirectComponent applicationDidOpenURL:url];'
    );
    return config;
  });
  return appDelegate;
};

const withAdyen = (config: ExpoConfig) => {
  return withAdyenIos(withAdyenAndroid(config));
};

export default withAdyen(config);
```

3. Generate a build with `eas build --profile development --platform ios`

And it should be working normally. If not please run `expo prebuild` to generate native files and double-check if these plugin functions are doing their job.

**Comment by descorp** [comment]
Great thanks @anthlasserre 💚
We will verify this and get back to you!

**Comment by bbialas** [comment]
<img width="420" alt="Screenshot 2023-09-21 at 14 52 31" src="https://github.com/Adyen/adyen-react-native/assets/1384286/72e3b78e-26d6-482f-b753-e5ccdec2200b">

I get an error in Expo managed App.

Any news about preparing expo modules for your library?

Thanks in advance!

**Comment by anthlasserre** [comment]
@descorp PR opened #288 

I created a plugin for the library and suggest building an example expo app in the project (in a next PR). 

Please let me know your thoughts.

**Comment by Seamus1989** [comment]
@descorp I have just done something very similar for v1.2.0. Hopefully this might help you! Could you give it a go and let me know!

```
const {
  withAppBuildGradle,
  withAppDelegate,
  withAndroidManifest,
  AndroidConfig,
  withMainActivity,
} = require('expo/config-plugins');



const withMyAdyenConfig = config => {
  // Edit AndroidManifest.xml with contents needed for Adyen
  config = withAndroidManifest(config, config => {
    const mainApplication = AndroidConfig.Manifest.getMainApplicationOrThrow(
      config.modResults,
    );
    if (!mainApplication.activity.length) {
      throw new Error('MainActivity not found');
    }
    const mainActivityInd = mainApplication.activity.findIndex(item => {
      return item.$['android:name'] === '.MainActivity';
    });
    if (mainActivityInd === -1) {
      throw new Error('MainActivity not found');
    }
    const intentFilterInd = mainApplication.activity[mainActivityInd][
      'intent-filter'
    ].findIndex(item => {
      // Please note with our current expo version. This <action android:name="android.intent.action.VIEW"/>
      // is present. And it is here we need to insert Adyen config.
      return item.action[0].$['android:name'] === 'android.intent.action.VIEW';
    });
    if (intentFilterInd === -1) {
      throw new Error('Intent filter not found');
    }

    mainApplication.activity[mainActivityInd]['intent-filter'].push({
      action: [{$: {'android:name': 'android.intent.action.VIEW'}}],
      category: [
        {$: {'android:name': 'android.intent.category.DEFAULT'}},
        {$: {'android:name': 'android.intent.category.BROWSABLE'}},
      ],
      data: [
        {
          $: {
            'android:scheme': '${redirectScheme}',
            'android:host': '${applicationId}',
          },
        },
      ],
    });

    mainApplication.service = {
      $: {
        'android:name':
          'com.adyenreactnativesdk.component.dropin.AdyenCheckoutService',
        'android:exported': 'false',
      },
    };

    return config;
  });

  // Edit app/build.gradle with contents needed for Adyen
  config = withAppBuildGradle(config, config => {
    // NOTE.
    // From Adyen/react-native "buildscript {" we see
    // rootProject.ext.adyenReactNativeRedirectScheme = "adyenreactnative"
    // if using manifestPlaceholders = [redirectScheme: rootProject.ext.adyenReactNativeRedirectScheme] we see an error during android build.
    config.modResults.contents = config.modResults.contents.replace(
      'defaultConfig {',
      'defaultConfig {\n        manifestPlaceholders = [redirectScheme: "adyenreactnative"]',
    );
    return config;
  });

  config = withMainActivity(config, config => {
    const mainActivity = config.modResults;
    mainActivity.contents = mainActivity.contents.replace(
      'import android.os.Bundle;',
      'import android.os.Bundle;\nimport com.adyenreactnativesdk.AdyenCheckout;\nimport android.content.Intent;',
    );
    mainActivity.contents = mainActivity.contents.replace(
      'super.onCreate(null);',
      'super.onCreate(null);\n      AdyenCheckout.setLauncherActivity(this);\n\n     }\n    @Override\n        public void onNewIntent(Intent intent) {\n            super.onNewIntent(intent);\n            AdyenCheckout.handleIntent(intent);\n    }\n    @Override\n    public void onActivityResult(int requestCode, int resultCode, Intent data) {\n      super.onActivityResult(requestCode, resultCode, data);\n      AdyenCheckout.handleActivityResult(requestCode, resultCode, data);',
    );
    return config;
  });

  // Edit AppDelegate.m with contents needed for Adyen.
  config = withAppDelegate(config, config => {
    const appDelegate = config.modResults;
    appDelegate.contents = appDelegate.contents.replace(
      '#import "AppDelegate.h"',
      '#import "AppDelegate.h"\n#import <adyen-react-native/ADYRedirectComponent.h>',
    );
    appDelegate.contents = appDelegate.contents.replace(
      'return [super application:application openURL:url options:options] || [RCTLinkingManager application:application openURL:url options:options];',
      'return  [ADYRedirectComponent applicationDidOpenURL:url] || [super application:application openURL:url options:options] || [RCTLinkingManager application:application openURL:url options:options];',
    );
    return config;
  });

  return config;
};

module.exports = withMyAdyenConfig;

```

`app.config.js`

```
const config = {
  expo: {
    name: "Example",
    ...
    plugins: [ ["./adyen.plugin.js"]],
    ...
```

Again, please make sure to run `expo prebuild --clean` prior to running any build commands.

**Comment by anthlasserre** [comment]
Hey @Seamus1989 feel free to check and try this opened PR 👇🏼 
https://github.com/Adyen/adyen-react-native/pull/288

And feel free to make any suggestions there.

**Comment by Seamus1989** [comment]
Nice @anthlasserre I'll pop this in there then (I'm actually just tidying up / testing), I don't suspect its perfect but with some other thoughts/feelings it might well help some people! Thanks

**Comment by camil-adyen** [comment]
We released Expo support in 2.0.0-beta3. I'll close this ticket for now.
